### DISCUSSION TITLE
Support for installation of additional python modules on nicegui docker start

### DISCUSSION THREAD

**Discussion by zan73** [zauberzeug/nicegui#3438] [Ideas / Feature Requests]
Potentially execute pip install -r requirements.txt if /app/requirements.txt exists, as part of docker startup

**Comment by rodja** [comment]
Do you have examples of popular images doing something similar? The "normal" way is to create your own Dockerfile and use it as a base to install the packages yourself. Like

```dockerfile
FROM zauberzeug/nicegui:1.4.31

COPY requirements.txt .

RUN pip install -r requirements.txt
```

**Reply by zan73** [reply]
I use linuxserver.io docker containers which allow for mapping of a separate volume for custom initialization, but that seems far too complex for this container.

I would imagine that other nicegui users would have a similar need (I myself need pandas and duckdb added to the base image).
I was hoping that something simple like a check for existence of /app/requirements.txt and then running pip would be a nice simple implementation but if you'd prefer not to change your container to cater for other python modules, happy to pursue another approach.

**Reply by rodja** [reply]
Ok. But as I understand it linuxserver.io's config system is also mainly used for configuration, not installing dependencies at startup. The disadvantages of installing packages at startup are:

- Slower startup and higher initial resource consumption
- Reduced reliability due to network dependencies and potential conflicts
- Decreased reproducibility and consistency across environments
- Security vulnerabilities from runtime package installations
- Inefficiency for multiple containers and scaling

But I understand that for a simple tryout or users with very little docker knowledge it would be "nice" to not write a Dockerfile at all. Maybe it would be a cool side project to provide a `nicegui-auto-install` image or something similar...

**Comment by zan73** [comment]
How about if we drive this via an environment variable. That way the end-user opts-in to the use of a requirements file.

DockerFile change:
```dockerfile
ENV REQUIREMENTS_FILE=""
```
And add this to /resources/docker-entrypoint.sh:

```sh
#Install Python modules from a requirements file passed as ENV variable to Docker
if [ -n "$REQUIREMENTS_FILE" ]; then \
    echo "Installing Python modules from $REQUIREMENTS_FILE"; \
    if ! pip install -r "$REQUIREMENTS_FILE"; then \
        echo "ERROR: Failed to install Python modules from $REQUIREMENTS_FILE" >&2; \
        exit 1; \
    fi; \
fi
```

**Reply by rodja** [reply]
Yes. That looks ok. But still I'm hesitant to put this into the official NiceGUI image due to the reasons listed in https://github.com/zauberzeug/nicegui/discussions/3438#discussioncomment-10293471.

**Reply by zan73** [reply]
OK, I've got it in my own git repo (https://github.com/zan73/nicegui), but I don't have a Docker Pro subscription for automated builds for when you push a new base image. Would you be willing to create a nicegui-auto-install as you suggested if you can do automated docker builds?

**Reply by rodja** [reply]
I would rather like it to be an external project. I just created a first draft PR to allow your project to build/publish the image on it's own: https://github.com/zan73/nicegui/pull/1.
Let's discuss it over there.
